codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2ConnectionHandlerBuilder.java

```diff
@@ -0,0 +1,399 @@
+     * would otherwise be exceeded.
+     */
+    protected B encoderEnforceMaxConcurrentStreams(boolean encoderEnforceMaxConcurrentStreams) {
+        enforceNonCodecConstraints("encoderEnforceMaxConcurrentStreams");
+        this.encoderEnforceMaxConcurrentStreams = encoderEnforceMaxConcurrentStreams;
+        return self();
+    }
+
+    /**
+     * Returns the {@link SensitivityDetector} to use.
+     */
+    protected SensitivityDetector headerSensitivityDetector() {
+        return headerSensitivityDetector != null ? headerSensitivityDetector : DEFAULT_HEADER_SENSITIVITY_DETECTOR;
+    }
+
+    /**
+     * Sets the {@link SensitivityDetector} to use.
+     */
+    protected B headerSensitivityDetector(SensitivityDetector headerSensitivityDetector) {
+        enforceNonCodecConstraints("headerSensitivityDetector");
+        this.headerSensitivityDetector = checkNotNull(headerSensitivityDetector, "headerSensitivityDetector");
+        return self();
+    }
+
+    /**
+     * Create a new {@link Http2ConnectionHandler}.
+     */
+    protected T build() {
+        if (encoder != null) {
+            assert decoder != null;
+            return buildFromCodec(decoder, encoder);
+        }
+
+        Http2Connection connection = this.connection;
+        if (connection == null) {
+            connection = new DefaultHttp2Connection(isServer());
+        }
+
+        return buildFromConnection(connection);
+    }
+
+    private T buildFromConnection(Http2Connection connection) {
+        Http2FrameReader reader = new DefaultHttp2FrameReader(validateHeaders);
+        Http2FrameWriter writer = new DefaultHttp2FrameWriter(headerSensitivityDetector());
+
+        if (frameLogger != null) {
+            reader = new Http2InboundFrameLogger(reader, frameLogger);
+            writer = new Http2OutboundFrameLogger(writer, frameLogger);
+        }
+
+        Http2ConnectionEncoder encoder = new DefaultHttp2ConnectionEncoder(connection, writer);
+        boolean encoderEnforceMaxConcurrentStreams = encoderEnforceMaxConcurrentStreams();
+
+        if (encoderEnforceMaxConcurrentStreams) {
+            if (connection.isServer()) {
+                encoder.close();
+                reader.close();
+                throw new IllegalArgumentException(
+                        "encoderEnforceMaxConcurrentStreams: " + encoderEnforceMaxConcurrentStreams +
+                        " not supported for server");
+            }
+            encoder = new StreamBufferingEncoder(encoder);
+        }
+
+        Http2ConnectionDecoder decoder = new DefaultHttp2ConnectionDecoder(connection, encoder, reader);
+        return buildFromCodec(decoder, encoder);
+    }
+
+    private T buildFromCodec(Http2ConnectionDecoder decoder, Http2ConnectionEncoder encoder) {
+        final T handler;
+        boolean success = false;
+        try {
+            // Call the abstract build method
+            handler = build(decoder, encoder, initialSettings);
+            success = true;
```

Comment:
could we remove this variable, `catch (Throwable t)` below, and do the `close` cleanup in the catch block?